vllm/model_executor/layers/quantization/fbgemm_fp8.py

```diff
@@ -0,0 +1,115 @@
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "fbgemm_fp8"
+
+    @classmethod
+    def get_supported_act_dtypes(cls) -> List[torch.dtype]:
+        return [torch.bfloat16]
+
+    @classmethod
+    def get_min_capability(cls) -> int:
+        return 90
+
+    @classmethod
+    def get_config_filenames(cls) -> List[str]:
+        return []
+
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "FBGEMMFp8Config":
+        return cls()
+
+    def get_quant_method(
+            self, layer: torch.nn.Module) -> Optional["QuantizeMethodBase"]:
+        if isinstance(layer, LinearBase):
+            return FBGEMMFp8LinearMethod(self)
+        return None
+
+    def get_scaled_act_names(self) -> List[str]:
+        return []
+
+
+class FBGEMMFp8LinearMethod(LinearMethodBase):
+
+    def __init__(self, quant_config: FBGEMMFp8Config):
+        self.quant_config = quant_config
+
+    def create_weights(
+        self,
+        layer: torch.nn.Module,
+        input_size_per_partition: int,
+        output_partition_sizes: List[int],
+        input_size: int,
+        output_size: int,
+        params_dtype: torch.dtype,
+        **extra_weight_attrs,
+    ):
+        del input_size, output_size
+        output_size_per_partition = sum(output_partition_sizes)
+
+        layer.logical_widths = output_partition_sizes
+
+        layer.input_size_per_partition = input_size_per_partition
+        layer.output_size_per_partition = output_size_per_partition
+        layer.orig_dtype = params_dtype
+
+        # WEIGHT
+        weight = Parameter(torch.empty(output_size_per_partition,
+                                       input_size_per_partition,
+                                       dtype=torch.float8_e4m3fn),
+                           requires_grad=False)
+        layer.register_parameter("weight", weight)
+        set_weight_attrs(weight, {
+            "input_dim": 1,
+            "output_dim": 0,
+            **extra_weight_attrs,
+        })
+
+        # WEIGHT SCALE
+        weight_scale = create_per_channel_scale_param(output_partition_sizes,
+                                                      **extra_weight_attrs)
+        layer.register_parameter("weight_scale", weight_scale)
+
+        # NOT USED FOR INFERNECE
+        input_scale_ub = torch.nn.Parameter(
+            torch.zeros((1), dtype=torch.float8_e4m3fn))
+        layer.register_parameter("input_scale_ub", input_scale_ub)
+        set_weight_attrs(input_scale_ub, {
+            "ignore_warning": True,
+            **extra_weight_attrs
+        })
+
+    def process_weights_after_loading(self, layer: Module) -> None:
+        weight = layer.weight
+        layer.weight = Parameter(weight.t(), requires_grad=False)
+
+    def apply(self,
+              layer: torch.nn.Module,
+              x: torch.Tensor,
+              bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+
+        return apply_fp8_linear(input=x,
```

Comment:
I don't think that apply_fp8_linear would work here. Using fbgemm it would look like:

            xq, x_scale = torch.ops.fbgemm.quantize_fp8_per_row(
                x, num_tokens, self.activation_scale_ub
            )
            y = torch.ops.fbgemm.f8f8bf16_rowwise(
                xq, layer.weight, x_scale, layer.weight_scale, use_fast_accum=True
            )

Particularly input_scale=None is most likely wrong, here is a reference implementation for quantize_fp8_per_row

def fp8_row_quantize_ref(x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
    # Quantize an input tensor and return the fp8 tensor and its inverse scale.
    x_row_max = torch.max(torch.abs(x), dim=1).values
    max_scaling_factor = E4M3_MAX_POS * 512.0  # Match kernel logics
    scale = torch.Tensor(E4M3_MAX_POS / x_row_max).clamp(max=max_scaling_factor)
    xq = (x * scale.unsqueeze(1)).to(fp8_e4m3)
    return xq, scale.reciprocal().to(torch.float32)

Reply:
yup, we are updating the quant kernel right now to use the `activation_scale_ub`

Reply:
@varun-sundar-rabindranath

Reply:
in combo with https://github.com/vllm-project/vllm/pull/6547/files

which enables per token scales

Reply:
Besides both torch._scaled_mm and ops.scaled_fp8_quant expect scale to be scalar

Reply:
Then it will match. `input_scale=None` in this function signals to use dynamic quantization rather than static

Reply:
#6547 extends `ops.scaled_fp8_quant` to accepts per token scales (vector of scales)

Reply:
`torch._scaled_mm` will not be used.

`cutlass_scaled_mm` accepts per channel weights and per row activations

Reply:
I pulled in #6547 to this PR so you can see

Reply:
Adds https://github.com/vllm-project/vllm/pull/6593/ adds the ub

Reply:
Okay, this PR currently now has the 
- `scale_ub`
- uses dynamic per token activation scales